track_syscalls/main.go

```diff
@@ -17,129 +20,170 @@ import (
+}
 
+func setupEBPF() (*ebpfObjects, *ringbuf.Reader, *link.Link, error) {
 	// Allow the current process to lock memory for eBPF resources.
 	if err := rlimit.RemoveMemlock(); err != nil {
-		log.Fatal(err)
+		return nil, nil, nil, fmt.Errorf("removing memlock: %w", err)
 	}
 
 	// Load pre-compiled eBPF program and maps into the kernel.
 	objs := ebpfObjects{}
 	if err := loadEbpfObjects(&objs, nil); err != nil {
-		log.Fatalf("loading objects: %v", err)
+		return nil, nil, nil, fmt.Errorf("loading objects: %w", err)
 	}
-	defer objs.Close()
 
 	// Open a tracepoint and attach the pre-compiled program.
 	tp, err := link.Tracepoint("raw_syscalls", "sys_enter", objs.TraceSyscall, nil)
 	if err != nil {
-		log.Fatalf("opening tracepoint: %s", err)
+		objs.Close()
+		return nil, nil, nil, fmt.Errorf("opening tracepoint: %w", err)
 	}
-	defer tp.Close()
 
-	// Open a ringbuf reader from userspace RINGBUF map described in the
-	// eBPF C program.
+	// Open a ringbuf reader from userspace RINGBUF map described in the eBPF C program.
 	rd, err := ringbuf.NewReader(objs.Events)
 	if err != nil {
-		log.Fatalf("opening ringbuf reader: %s", err)
+		tp.Close()
+		objs.Close()
+		return nil, nil, nil, fmt.Errorf("opening ringbuf reader: %w", err)
 	}
-	defer rd.Close()
-
-	// Close the reader when the process receives a signal, which will exit
-	// the read loop.
-	go func() {
-		<-stopper
-
-		if err := rd.Close(); err != nil {
-			log.Fatalf("closing ringbuf reader: %s", err)
-		}
-	}()
 
-	log.Println("Waiting for events..")
+	return &objs, rd, &tp, nil
+}
 
-	// bpfEvent is generated by bpf2go.
-	var event ebpfEvent
-	for {
-		record, err := rd.Read()
-		if err != nil {
-			if errors.Is(err, ringbuf.ErrClosed) {
-				log.Println("Received signal, exiting..")
-				return
+func runBuildMode(binaryPath string) {
+	syscalls := make(map[string]map[int]bool)
+	setupAndRun(binaryPath, func(event ebpfEvent, stackTrace []uint64, objs *ebpfObjects) {
+		callerPackage := stackanalyzer.GetCallerPackage(stackTrace)
+		if callerPackage != "" {
+			if _, ok := syscalls[callerPackage]; !ok {
+				syscalls[callerPackage] = make(map[int]bool)
 			}
-			log.Printf("reading from reader: %s", err)
-			continue
+			syscalls[callerPackage][int(event.Syscall)] = true
 		}
+		logEvent(event, stackTrace, objs)
+	})
+	if err := syscallfilter.Write(syscallfilter.ConvertSyscallsMap(syscalls)); err != nil {
+		log.Fatalf("Writing allowlist JSON: %v", err)
+	}
+	log.Println("Build mode completed. Allowlist JSON file created.")
+}
 
-		// Parse the ringbuf event entry into a bpfEvent structure.
-		if err := binary.Read(bytes.NewBuffer(record.RawSample), binary.LittleEndian, &event); err != nil {
-			log.Printf("parsing ringbuf event: %s", err)
-			continue
-		}
+func runTraceMode(binaryPath string) {
+	setupAndRun(binaryPath, func(event ebpfEvent, stackTrace []uint64, objs *ebpfObjects) {
+		// Just log the event
+		logEvent(event, stackTrace, objs)
```

Comment:
We would only want to log the violation of the event, eventually right? This would log complete stacktrace based on all eBPF comments, right?

Reply:
For now, for each execution mode, I am logging  everything for debugging and testing purpose. We can eventually add a debug mode flag for this. 

But yes, in general, we want to log only violation events for the enforcement mode. 
We will log everything, as logEvent does, for a "monitoring" mode (now called 'trace'), if we stille need this mode.